snapcraft/internal/elf.py

```diff
@@ -169,6 +170,58 @@ def get_elf_files(root: str, file_list: Sequence[str]) -> FrozenSet[str]:
         # for an ldd call.
         file_m = ms.file(path_b)
         if file_m.startswith('ELF') and 'dynamically linked' in file_m:
-            elf_files.add(path)
-
-    return frozenset(elf_files)
+            is_executable = 'interpreter' in file_m
+            elf_files[path] = dict(executable=is_executable)
+
+    return elf_files
+
+
+class Patcher:
+    """Patcher holds the necessary logic to patch elf files."""
+
+    def __init__(self, *, dynamic_linker: str) -> None:
+        """Create a Patcher instance.
+
+        :param str dynamic_linker: the path to the dynamic linker to set the
+                                   elf file to.
+        """
+        self._dynamic_linker = dynamic_linker
+
+        # If we are running from the snap we want to use the patchelf
+        # bundled there as it would have the capabilty of working
+        # anywhere given the fixed ld it would have.
+        # If not found, resort to whatever is on the system brought
+        # in by packaging dependencies.
+        if common.is_snap():
+            snap_dir = os.getenv('SNAP')
+            self._patchelf = os.path.join(snap_dir, 'bin', 'patchelf')
+        else:
+            self._patchelf = 'patchelf'
+
+    def patch(self, *, elf_file: str, elf_properties: Dict[str, str]) -> None:
+        """Patch elf_file with the Patcher instance configuration.
+
+        patch will to the right thing with regards to the different type
+        of elf files.
+
+        :param str elf_file: path to the elf_file to patch.
+        :param dict elf_properties:
+            properties of the elf file. The property list is as follows:
+                * executable: boolean stating if elf_file is executable.
```

Comment:
Yeah if we're making it part of the API, I doubly think this should be a class. It'll make the API much less error-prone.

Reply:
Loving the new class, though this bit of the docstring needs to be updated to reflect.